Azure Object Anchors data plane SDK (#18871)
Azure/azure-sdk-for-net: sdk/objectanchors/Azure.MixedReality.ObjectAnchors/README.md

# Azure Object Anchors client library for .NET

> Azure Object Anchors enables an application to detect an object in the physical world using a 3D model and estimate its 6-DoF pose. This package supports the conversion of an existing 3D asset into a form that can be used by the Object Anchors runtime to detect physical objects.

Author: @craigktreasure 
Adopted suggested language
RE: "I might suggest stealing some of the language from the docs.", and the suggested change

    existing Azure AD security groups. You can also grant access directly to users in your organization.
  - Otherwise, we recommend that you obtain Azure AD tokens from a web service that supports your app. We recommend this
    method for production applications because it allows you to avoid embedding the credentials for access to the Object Anchors asset conversion service in your client application.

See [here](https://docs.microsoft.com/azure/spatial-anchors/concepts/authentication) for detailed instructions and information.

## Key concepts

### ObjectAnchorsClient

The `ObjectAnchorsClient` is the client library used to access the Object Anchors asset conversion service. From there, a storage upload URI will be provided for users to upload their assets for conversion into a format that is usable by the Object Anchors runtime.

Assets uploaded to the Microsoft-hosted URI obtained from the client will be retained for **48 hours**.

The final converted model in Microsoft-hosted storage will be retained for **48 hours**. 

## Examples

### Upload an asset for Object Anchors asset conversion

```csharp

> TokenCredential credential = new DefaultAzureCredential(true);

Reviewer: >DefaultAzureCredential [](start = 33, length = 22)

I'd prefer we instead document this using the `AzureKeyCredential`. It's the easiest way for people to quickly get started and what we'll be using in the quickstarts.

Author: Pending

Author: Replaced with `AzureKeyCredential`

ObjectAnchorsClient client = new ObjectAnchorsClient(account, credential);

> Uri uploadedInputAssetUri = (await client.GetAssetUploadUriAsync()).Value.InputAssetUploadUri;

Reviewer: ```suggestion
GetAssetUploadUriResult assetUploadUriResponse = await client.GetAssetUploadUriAsync();
Uri uploadedInputAssetUri = assetUploadUriResponse.InputAssetUploadUri;
```

Author: Pending

BlobClient blobClient = new BlobClient(uploadedInputAssetUri);

using (FileStream fs = File.OpenRead(localFilePath))
{
  await blobClient.UploadAsync(fs);
}
```

> ### Begin Object Anchors asset conversion after uploading an asset

Reviewer: ```suggestion
### Start 3D asset conversion

After uploading the 3D asset, you'll need to start a conversion job.
```

Author: Pending

```csharp
StartAssetConversionOptions ingestionJobOptions = new StartAssetConversionOptions(uploadedInputAssetUri, AssetFileType.FromFilePath(localFilePath), assetGravity, scale);

> AssetConversionOperation operation = await client.StartAssetConversionAsync(ingestionJobOptions);

Reviewer: ```suggestion
StartAssetConversionOptions assetConversionOptions = new StartAssetConversionOptions(uploadedInputAssetUri, AssetFileType.FromFilePath(localFilePath), assetGravity, scale);

AssetConversionOperation operation = await client.StartAssetConversionAsync(assetConversionOptions);
```

The term "ingestion" doesn't mean much to customers. I'd go with either "assetConversion" or "conversion".

Author: Pending

Author: Ingestion replaced with assetconversion
